pkg/controller/buildrun/runtime_image.go

```diff
@@ -0,0 +1,184 @@
+LABEL {{ $k }}="{{ $v }}"
+{{- end }}
+
+{{- range $dir := .Spec.Runtime.Directories }}
+{{- $parts := splitDirectories $dir }}
+COPY --from=builder "{{ index $parts 0 }}" "{{ index $parts 1 }}"
+{{- end }}
+
+{{- if .Spec.Runtime.WorkDir }}
+WORKDIR "{{ .Spec.Runtime.WorkDir }}"
+{{- end }}
+
+{{- if .Spec.Runtime.Entrypoint }}
+ENTRYPOINT [ \
+{{- range $c := .Spec.Runtime.Entrypoint }}
+	"{{ $c }}", \
+{{- end }}
+]
+{{- end -}}
+`
+
+	// workspaceDir common workspace directory, where the source code is located.
+	workspaceDir = "/workspace/source"
+
+	// buildahStable stable buildah image.
+	buildahStable = "quay.io/buildah/stable:latest"
+
+	// runtimeDockerfile runtime Dockerfile file name.
+	runtimeDockerfile = "Dockerfile.runtime"
+)
+
+// runtimeDockerfilePath path to runtime Dockerfile on workspace directory.
+var runtimeDockerfilePath = path.Join(workspaceDir, runtimeDockerfile)
+
+// splitDirectories split informed directory by colon, returning a slice with parts. When colon is
+// not present, return the informed directory twice. This method always returns a slice with two
+// entries.
+func splitDirectories(dir string) []string {
+	parts := strings.Split(dir, ":")
+	if len(parts) == 2 {
+		return parts
+	}
+	return []string{dir, dir}
+}
+
+// renderRuntimeDockerfile render runtime Dockerfile using build instance and pre-defined template.
+func renderRuntimeDockerfile(b *buildv1alpha1.Build) (*bytes.Buffer, error) {
+	tmpl, err := template.New(runtimeDockerfile).
+		Funcs(template.FuncMap{
+			"splitDirectories": splitDirectories,
+		}).
+		Parse(runtimeDockerfileTmpl)
+	if err != nil {
+		return nil, err
+	}
+
+	dockerfile := new(bytes.Buffer)
+	if err = tmpl.Execute(dockerfile, b); err != nil {
+		return nil, err
+	}
+	return dockerfile, nil
+}
+
+// runtimeDockerfileStep trigger the rendering of Dockerfile.runtime, and use this input as a
+// build-step to create a new file.
+func runtimeDockerfileStep(b *buildv1alpha1.Build) (*buildv1alpha1.BuildStep, error) {
+	dockerfile, err := renderRuntimeDockerfile(b)
+	if err != nil {
+		return nil, err
+	}
+
+	rootUserID := int64(0)
+	container := v1.Container{
+		Image: b.Spec.BuilderImage.ImageURL,
+		SecurityContext: &v1.SecurityContext{
+			RunAsUser: &rootUserID,
+		},
+		WorkingDir: workspaceDir,
+		Command:    []string{"/bin/bash"},
+		Args: []string{
+			"-x",
+			"-c",
+			fmt.Sprintf("echo '%s' >%s", dockerfile, runtimeDockerfilePath),
+		},
+	}
+	return &buildv1alpha1.BuildStep{Container: container}, nil
+}
+
+// runtimeBuildStep returns a build-step to build the Dockerfile.runtime with buildah.
+func runtimeBuildStep(b *buildv1alpha1.Build) *buildv1alpha1.BuildStep {
+	contextDir := "."
+	if b != nil && b.Spec.Source.ContextDir != nil {
+		contextDir = *b.Spec.Source.ContextDir
+	}
+
+	truePtr := true
+	container := v1.Container{
+		Image:      buildahStable,
+		WorkingDir: workspaceDir,
+		SecurityContext: &v1.SecurityContext{
+			Privileged: &truePtr,
```

Comment:
Can this step run without buildah or Privileged?
`Privileged=true` cannot work on our side now

Reply:
Should we use Kaniko for this step then?

Reply:
I think `buildah` does run in non-privileged container, however, I need to do some "homework" on this subject to be honest. (cc @gabemontero)

Reply:
On latest commit we are using `kaniko` for the runtime-image build.